File: packages/solver-wasm/src/puyo_attr.rs
use num_derive::{FromPrimitive, ToPrimitive};
use num_traits::ToPrimitive;

#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq, FromPrimitive, ToPrimitive)]
pub enum PuyoAttr {
    Red = 1,
    Blue = 2,
    Green = 3,
    Yellow = 4,
    Purple = 5,
    Heart = 6,
    Prism = 7,
    Ojama = 8,
    Kata = 9,
    Padding = 10,
}

pub const COLOR_ATTRS: [PuyoAttr; 5] = [
    PuyoAttr::Red,
    PuyoAttr::Blue,
    PuyoAttr::Green,
    PuyoAttr::Yellow,
    PuyoAttr::Purple,
];
pub const SPECIAL_ATTRS: [PuyoAttr; 4] = [
    PuyoAttr::Heart,
    PuyoAttr::Prism,
    PuyoAttr::Ojama,
    PuyoAttr::Kata,
];

pub const POPPABLE_ATTRS: [PuyoAttr; 8] = [
    PuyoAttr::Red,
    PuyoAttr::Blue,
    PuyoAttr::Green,
    PuyoAttr::Yellow,
    PuyoAttr::Purple,
    PuyoAttr::Heart,
    PuyoAttr::Prism,
    PuyoAttr::Ojama,
];

pub fn is_colored_attr(attr: PuyoAttr) -> bool {
    if attr.to_usize() >= PuyoAttr::Red.to_usize() && attr.to_usize() <= PuyoAttr::Purple.to_usize()
    {
        true
    } else {
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_colored_attr() {
        assert_eq!(is_colored_attr(PuyoAttr::Red), true);
        assert_eq!(is_colored_attr(PuyoAttr::Blue), true);
        assert_eq!(is_colored_attr(PuyoAttr::Green), true);
        assert_eq!(is_colored_attr(PuyoAttr::Yellow), true);
        assert_eq!(is_colored_attr(PuyoAttr::Purple), true);
        assert_eq!(is_colored_attr(PuyoAttr::Heart), false);
        assert_eq!(is_colored_attr(PuyoAttr::Prism), false);
        assert_eq!(is_colored_attr(PuyoAttr::Ojama), false);
        assert_eq!(is_colored_attr(PuyoAttr::Kata), false);
        assert_eq!(is_colored_attr(PuyoAttr::Padding), false);
    }
}

File: src/logics/PuyoAttr.ts
/** ぷよの属性 (色ぷよ以外の特殊なぷよも含む) */
export enum PuyoAttr {
  /** 赤 */
  Red = 1,
  /** 青 */
  Blue = 2,
  /** 緑 */
  Green = 3,
  /** 黄 */
  Yellow = 4,
  /** 紫 */
  Purple = 5,
  /** ハート */
  Heart = 6,
  /** プリズム */
  Prism = 7,
  /** おじゃま */
  Ojama = 8,
  /** 固 */
  Kata = 9,
  /** パディング */
  Padding = 10
}

/** 色付きぷよの属性 */
export type ColoredPuyoAttr =
  | PuyoAttr.Red
  | PuyoAttr.Blue
  | PuyoAttr.Green
  | PuyoAttr.Yellow
  | PuyoAttr.Purple;

const puyoAttrMap: ReadonlyMap<PuyoAttr, string> = new Map([
  [PuyoAttr.Red, '赤'],
  [PuyoAttr.Blue, '青'],
  [PuyoAttr.Green, '緑'],
  [PuyoAttr.Yellow, '黄'],
  [PuyoAttr.Purple, '紫'],
  [PuyoAttr.Heart, 'ハート'],
  [PuyoAttr.Prism, 'プリズム'],
  [PuyoAttr.Ojama, 'おじゃま'],
  [PuyoAttr.Kata, '固']
]);

/** 色ぷよの属性リスト */
export const coloredPuyoAttrList: ReadonlyArray<ColoredPuyoAttr> = [
  PuyoAttr.Red,
  PuyoAttr.Blue,
  PuyoAttr.Green,
  PuyoAttr.Yellow,
  PuyoAttr.Purple
];

/**
 * ぷよの属性名を取得する
 * @param puyoAttribute
 * @returns
 */
export const getPuyoAttrName = (
  puyoAttribute: PuyoAttr | undefined
): string => {
  return puyoAttrMap.get(puyoAttribute!) ?? '';
};

/**
 * 色ぷよかどうか (プリズムは含まない)
 * @param puyoAttr ぷよの属性
 * @returns
 */
export const isColoredPuyoAttr = (
  puyoAttr: PuyoAttr
): puyoAttr is ColoredPuyoAttr => {
  return puyoAttr >= PuyoAttr.Red && puyoAttr <= PuyoAttr.Purple;
};
